### ISSUE TITLE
Getting rid of cucumber

### ISSUE THREAD

**Issue by troessner** [troessner/reek#592]
I stopped using cucumber in commercial projects long time ago and every time one of our cucumber features fails I remember why:

The additional amount of indirection and abstraction only makes sense if you really have a non-technical person on board like a product manager that actually not only prefers, but needs the feature description in Gherkin style.
But for us this doesnt make much sense.
Here's what happens when one of our features fails:
  [1] I look at the feature
  [2] Then I look at the source
  [3] I fix the source
  [4] And hope the feature passes now
  [5] If it doesnt I have a significantly higher effort to debug a separate cucumber process than an rspec one

[1] and [4] are completely unnecessary and [5] is a huge pain in the ass.

In a nutshell: AFAIS cucumber is not suited for and I suggest that we use rspec integration tests for all new features with the exception of those that rely heavily on the filesystem (for which we can use aruba).

WDYT?

**Comment by chastell** [comment]
Oh yes so much this please! Other than the debugging pains waiting for the features to pass is an exercise in patience. Let’s stick to RSpec for integration / acceptance and maybe slowly port the Cucumber features as well (I’ll be happy to do this in small steps now and then).

**Comment by mvz** [comment]
I have many thoughts on this, so this may be long.

First, moving our features from cucumber to RSpec probably won't make them much faster.

Second, I have yet to see an RSpec integration test that is as readable as Cucumber. You usually end up with the implementation of what would be each Cucumber step being inlined in the specs. On the other hand, if we force ourselves to turn each step into a method or oneliner, we can get something like this, which I think looks fine:

```ruby
  it 'runs reek succesfully' do
    create_demo_file 'demo.rb'
    run 'reek remo.rb'
    expect(exit_status).to indicate_smells
    expect(stdout).to eq <<-EOF
    demo.rb -- 8 warnings:
        [1]:Dirty has no descriptive comment (IrresponsibleModule)
        [3]:Dirty#awful has 4 parameters (LongParameterList)
        [3]:Dirty#awful has boolean parameter 'log' (BooleanParameter)
        [3]:Dirty#awful has the parameter name 'x' (UncommunicativeParameterName)
        [5]:Dirty#awful has the variable name 'w' (UncommunicativeVariableName)
        [3]:Dirty#awful has unused parameter 'log' (UnusedParameters)
        [3]:Dirty#awful has unused parameter 'offset' (UnusedParameters)
        [3]:Dirty#awful has unused parameter 'y' (UnusedParameters)
    EOF
```

Thirdly, yes, debugging a feature is harder than debugging a unit test, but in the example above you'll still be running a separate process so I'm not sure the problems will go away. Aruba should help out here a bit by keeping the temporary directory around if a feature fails, so your best bet is to run reek by hand if the feature fails.

And finally, I think we have too many integration tests, sometimes for silly things like combining command line options. We can probably move those tests to a more unit-like RSpec spec so we're left with just a few real integration tests that basically confirm that reek works: command-line, api, rake task, samples. Even the samples could be checked in a more focused test.

**Comment by troessner** [comment]
>> Second, I have yet to see an RSpec integration test that is as readable as Cucumber. You usually end up with the implementation of what would be each Cucumber step being inlined in the specs. On the other hand, if we force ourselves to turn each step into a method or oneliner, we can get something like this, which I think looks fine:

Agreed.

>> And finally, I think we have too many integration tests, sometimes for silly things like combining command line options. We can probably move those tests to a more unit-like RSpec spec so we're left with just a few real integration tests that basically confirm that reek works: command-line, api, rake task, samples. Even the samples could be checked in a more focused test.

Sounds good, seems like a good compromise.

**Comment by troessner** [comment]
I feel this has been discussed enough, closing.
